### PR TITLE
Propagate user thrown errors out of IO routines, channels no longer store error codes

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR wraps up the overhaul of error handling for Chapel IO. It adjusts IO routines so that they correctly propagate user thrown errors out of `readThis` / `writeThis`.

Prior to this effort, when a user threw an error from `readThis`/`writeThis` routine, it would be caught by IO machinery and an `EIO` error would be placed in the channel as a placeholder. In the new world, _errors are not stored in channels at all_. A user thrown error will now propagate all the way out of top level IO routines.

In the old world, code might have made use of the `<~>` operator in conjunction with the `channel.getError` routine to write conditional reads:

```chapel
proc myRecord.readThis(f) throws {
  f <~> new ioLiteral("optional prefix");
  if f.error() == EFORMAT {
    // We didn't find the prefix.
    f.clearError();
  } else {
    // Read whatever we expect to follow the prefix.
  }

  f <~> this.foo;
  f <~> this.bar;
  // ...
}
```

Channels no longer store error codes. Instead, all errors are thrown to the user directly (in this case, thrown from `<~>`). Here's the above code in the new world:

```chapel

proc myRecord.readThis(f) throws {
  try {
    f <~> new ioLiteral("optional prefix");
    // Read whatever we expect to follow the prefix.
  } catch err: BadFormatError {
    // We didn't find the prefix. Nothing to do here now.
  }

  f <~> this.foo;
  f <~> this.bar;
  // ...
}
```

Errors thrown from `readThis`/`writeThis` by the user will now correctly propagate out of IO code instead of being caught and replaced by an `IOError`. This makes it possible for users to throw more descriptive errors when overloading `readThis`/`writeThis`.

Because channels no longer store error codes, distinguishing between `SystemError` thrown by the user and `SystemError` thrown by IO machinery no longer matters, except in the case of `channel.read` (which returns `false` instead of throwing a `EndOfFileError`). 

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none`
- [ ] `ALL` on `linux64` when `CHPL_COMM=gasnet`

### CODE REVIEW

File: modules/internal/DefaultAssociative.chpl [hunk 1]
```diff
@@ -150,35 +150,19 @@ module DefaultAssociative {
...
-              f.clearError();
-            } else {
-              // Stop reading if we got to the end
-              // or if there was another error.
+            // Try reading an end curly. If we get it, break.
+            try {
+              f <~> end;
               break;
-            }
+            } catch err: BadFormatError {}
```
Comment (mppf) [inline modules/internal/DefaultAssociative.chpl]: Generally speaking, a `catch { }` block is really bad style. I'd appreciate it very much if in this PR you could put a comment inside each of these to say something like "Didn't read end, so forget we tried, and keep reading".

File: modules/standard/IO.chpl [hunk 2]
```diff
@@ -3010,19 +3010,59 @@ private inline proc _write_binary_internal(_channel_internal:qio_channel_ptr_t,
...
+
+  return result;
+}
+
+//
+// The channel must be locked and running on this.home.
+// The intent of x is ref (vs out) because it might contain a string literal.
+//
+pragma "no doc"
+proc channel._readOne(param kind: iokind, ref x:?t,
```
Comment (mppf) [inline modules/standard/IO.chpl]: Let's put a comment like "TODO: incorporate this logic into _read_one_internal and make it a method"

File: modules/standard/IO.chpl [hunk 3]
```diff
@@ -3091,42 +3131,24 @@ private inline proc _read_one_internal(_channel_internal:qio_channel_ptr_t,
...
-  //
-  } catch err: SystemError {
-    const code = err.err;
-    _qio_channel_set_error_unlocked(_channel_internal, code);
-  } catch err {
-    const code: syserr = EIO;
-    _qio_channel_set_error_unlocked(_channel_internal, code);
-  }
+  // Code in defer blocks fires after deinitializer, see #14797.
+  ref r = reader;
```
Comment (mppf) [inline modules/standard/IO.chpl]: If you rebase, you should be able to remove these workarounds.

File: modules/standard/IO.chpl [hunk 4]
```diff
@@ -6643,17 +6603,25 @@ proc channel.readf(fmtStr:?t, ref args ...?k): bool throws
...
-    this._set_style(save_style);
   }
 
   if !err {
     return true;
   } else if err == EEOF {
+    //
+    // TODO: Shouldn't we also return false for EFORMAT? Yet some tests
+    // (io/ferguson/json.chpl off the top of my head) expect EFORMAT errors
+    // to be thrown. Is the documentation off?
```
Comment (mppf) [inline modules/standard/IO.chpl]: What made you think it should return false for EFORMAT? I think I'd expect it to throw.

Comment (dlongnecke-cray) [reply modules/standard/IO.chpl]: The documentation for `channel.readf` says that it "false on EEOF or if the format did not match the input". Should the docstring be updated instead?

Comment (mppf) [reply modules/standard/IO.chpl]: I don't know what to do with it. Maybe make an issue about it?

Comment (dlongnecke-cray) [reply modules/standard/IO.chpl]: Sure, I can do that.

**Comment by dlongnecke-cray** [comment]
One more test failure discovered in testing for `CHPL_COMM=gasnet` (that I somehow missed the first time?!), which is `test/performance/hilde/userSync`.

**Pull request merged**